### DISCUSSION TITLE
Loading components as raw string.

### DISCUSSION THREAD

**Discussion by anubra266** [remix-run/remix#3399] [Q&A] [answered]
I can import a text file.
```js
import textContent from "~/components/preview/alerts/success.txt";
```
If I import a custom extension
```js
import rr from "~/components/preview/alerts/success.choc";
```
I get
<img width="605" alt="CleanShot 2022-06-06 at 13 52 31@2x" src="https://user-images.githubusercontent.com/30869823/172164295-fb693a4f-9f6c-454f-80d8-8670c1408d2e.png">

I'm interpreting that as there's a way to configure file loaders in remix. And by default there's a loader for text files.

What I'm trying to do is import a component as text. i.e. I get the code content of the file as a string literal.
In NextJS I used to use webpack raw loader, something like:

```js
const preCode = require(`!!raw-loader!pages/preview/comp.tsx`).default;
```
And get the content of that file as text.

I want to know if this is possible in Remix or If I can configure the loader for something like this myself.

I was switching a project from nextjs to remix, It's a big one and I was nearly done, before I encountered this snag. It's too late to give up now. 😢 

Thanks in advance.

**Comment by kiliman** [comment]
See this answer for how to patch Remix and add support for your custom extension.

https://github.com/remix-run/remix/discussions/2468#discussioncomment-2424868

**Reply by kiliman** [reply]
BTW Remix doesn't process your imports. I simply returns the URL of the hashed asset. Remix copies the file to `/public/build/_assets`.

If you need the content, you'll need to read the file on the server.

**Reply by anubra266** [reply]
Thanks @kiliman This helps with exporting other file extensions, but what's the possibility on exporting a component as a string?

**Reply by anubra266** [reply]
> BTW Remix doesn't process your imports. I simply returns the URL of the hashed asset. Remix copies the file to `/public/build/_assets`.
> 
> If you need the content, you'll need to read the file on the server.

The content of the copied files in build is already parsed, i.e. has stuff like `React.createElement` I'm looking to get the content raw exactly how they are in the code.

**Reply by anubra266** [reply]
Also, @kiliman
![CleanShot 2022-06-06 at 16 00 45@2x](https://user-images.githubusercontent.com/30869823/172187568-9953a4d6-c6a7-43cd-8fbf-ee06c9993a2b.png)

But it's resolved in code

<img width="688" alt="CleanShot 2022-06-06 at 16 02 49@2x" src="https://user-images.githubusercontent.com/30869823/172187959-de5ec561-07cc-48f6-9741-2fbf1379e0aa.png">

**Reply by kiliman** [reply]
Did you restart remix dev after patching your file? Again, this is only going to give you the URL to the asset. If you need the content, you'll need to use something like `fs.readFile` in your loader as suggested by @sergiodxa.

If this isn't helping, perhaps a simple repo can help us unblock you.

**Reply by anubra266** [reply]
@kiliman I think it's because I'll have to add it here : https://github.com/remix-run/remix/blob/6a2bf52d9ca60e2ce6459ebe72df28e6a568ed26/packages/remix-dev/compiler/loaders.ts#L4

**Reply by anubra266** [reply]
@kiliman After some more research on the remix codebase, This is a matter of esbuild, I'll like to know, how can I extend esbuild configuration in remix?

**Comment by sergiodxa** [comment, marked as answer]
I think rather than depending on a loader in the compiler, you could read the file inside a loader using `fs.readFile`, then return the content from the loader to your component.

**Reply by IHIutch** [reply]
Do you have an example of this working or of what you're thinking here? I'm running into a similar issue trying to pull in the contents of a CSS as a string.

I have `import styles from '~/styles/mystyles.css'` which when I console.log I see `'/build/_assets/mystyles-A2DXJUKS.css'`

But when I try to do `const styleAsString = readFileSync(styles, 'utf-8')` I get an error saying `ENOENT: no such file or directory, open '/build/_assets/mystyles-A2DXJUKS.css'`

**Reply by IHIutch** [reply]
I run into the same issue when I try to use the server path, though interestingly, I've found that if I put a static file and path in the API directory like: `api/styles/mystyles.css` I'm able to access the contents by doing `readFileSync(__dirname + '/styles/mystyles.css', 'utf-8')`. But not when I `import` and try to use the generated path

**Reply by kiliman** [reply]
When you import an asset, Remix returns the URL not the file path. The asset is copied to your `public/builds` folder.

To get the actual file path, you'll want to do something like:

```ts
import styles from '~/styles/mystyles.css'
const filePath = './public' + styles    // convert /build/_assets to ./public/build/_assets
const css = fs.readFileSync(filePath, 'utf-8')
```

**Reply by IHIutch** [reply]
Ah, I see, tysm.

**Reply by IHIutch** [reply]
Hmm, actually `const filePath = './public' + styles` does not appear to be working in my production (it is locally, though) build, at least when deploying on Vercel.

My guess is that's because `/build` is the folder that's deployed?

**Reply by IHIutch** [reply]
Spent some time investigating this... It's a little finicky with Vercel. [This thread](https://github.com/remix-run/remix/discussions/2000) pointed me to the Remix discord which, if you have discord, you view the [related message](https://discord.com/channels/770287896669978684/778004202282287104/997966494296195163).

Otherwise, check out the [comment left here](https://github.com/theMosaad/remix-vercel-read-files/blob/f5542bc0cb7579f0ea0a311142703df0055e459c/app/routes/index.tsx#L10-L13) to come up with a working solution to reading static files on Vercel with remix.

So my final solution looks like this: 
```
const stylePath =
    process.env.NODE_ENV === 'production'
      ? __dirname + '/../public' + styles
      : './public' + styles
```

Fwiw, this ends up being the path on Vercel: `/var/task/api/../public/build/_assets/mystyles-A2DXJUKS.css`
Why that path works, I don't fully understand

(thanks @theMosaad)
